File: Workshop1/client-socket.py
# Client in Python

import logging
import random
from socket import *
import string
import threading

logging.basicConfig(
    level = logging.INFO,
    format = "%(asctime)s - %(levelname)s - %(message)s"
)

DEFAULT_NUMBER_OF_CLIENTS:int = 3
DEFAULT_SERVER_NAME:str = "172.23.197.112"
DEFAULT_PORT_NUMBER:int = 12000
MIN_PORT_NUMBER:int = 0
MAX_PORT_NUMBER:int = 65535
BYTES_DATA_CHUNK:int = 1024

def serverIdentification() -> tuple:
    serverName:str = input("Enter server hostname or IP address: ")
    if not serverName:
        serverName = DEFAULT_SERVER_NAME
        logging.info(f"No IP address specified. Using default address: {DEFAULT_SERVER_NAME}")
    serverPort:str = input("Enter server port number: ")
    if not serverPort:
        serverPort = DEFAULT_PORT_NUMBER
        logging.info(f"No port number specified. Using default port number: {DEFAULT_PORT_NUMBER}")
    try:
        serverPort_int:int = int(serverPort)
    except:
        serverPort_int = DEFAULT_PORT_NUMBER
        logging.info(f"Invalid port number. Using default port number: {DEFAULT_PORT_NUMBER}")
    if serverPort_int <= MIN_PORT_NUMBER or serverPort_int > MAX_PORT_NUMBER:
        serverPort_int = DEFAULT_PORT_NUMBER
    logging.info(f"Preparing to connect to the server...")
    return (serverName, serverPort_int)

def clientInitialization(clientID:int, serverName:str, serverPort:int) -> None:
    numberOfMessages = random.randint(1, 3)
    clientSocket:socket = socket(AF_INET, SOCK_STREAM)
    try:
        clientSocket.connect((serverName, serverPort))
    except Exception as exception:
        logging.error(f"Client {clientID} failed: {exception}")
        break # Break the loop if connection fails.
    for message in range(numberOfMessages):
        randomText = "".join(random.choices(string.ascii_lowercase, k = 8))
        sentence:str = f"Client {clientID}, message {message + 1}: {randomText}"
        logging.info(f"Client {clientID} successfully connected to the server.")
        clientSocket.send(sentence.encode())
        modifiedSentence:bytes = clientSocket.recv(BYTES_DATA_CHUNK)
        print(f"[Client {clientID}] Message from server:", modifiedSentence.decode())
        clientSocket.close()

def main() -> None:
    numberOfClients:str = input("Enter the number of clients: ")
    if not numberOfClients:
        numberOfClients_int = DEFAULT_NUMBER_OF_CLIENTS
        logging.info(f"No number of clients specified. Using {DEFAULT_NUMBER_OF_CLIENTS} clients.")
    else:
        numberOfClients_int = int(numberOfClients)
    serverName, serverPort = serverIdentification()
    threads:list = []
    for _id in range(numberOfClients_int):
        thread:threading.Thread = threading.Thread(target = clientInitialization, args = (_id + 1, serverName, serverPort))
        threads.append(thread)
        thread.start()

if __name__ == "__main__":
    main()

File: Workshop1/client-socket.c
// Client socket in C (Multithreaded & Automated)
#include <errno.h>
#include <netdb.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <sys/types.h>
#include <sys/socket.h>
#include <unistd.h>
#include <pthread.h>
#include <time.h>
#include <arpa/inet.h>
#include <netinet/in.h> // Asegura la correcta traducción de puertos

#define PORT 12000
#define NUM_CLIENTS 3

char server_ip[64];

typedef struct
{
    int client_id;
} thread_data_t;

void *client_thread(void *arg)
{
    thread_data_t *data = (thread_data_t *)arg;
    int client_id = data->client_id;
    free(data);

    struct sockaddr_in sa;
    int socket_fd;
    char buffer[BUFSIZ];

    memset(&sa, 0, sizeof sa);
    sa.sin_family = AF_INET; // IPv4
    sa.sin_port = htons(PORT);

    // NUEVO: Validación estricta y segura de la IP
    if (inet_pton(AF_INET, server_ip, &sa.sin_addr) <= 0)
    {
        fprintf(stderr, "[Client %d] Invalid IP address format: %s\n", client_id, server_ip);
        pthread_exit(NULL);
    }

    // Create socket, connect it to remote server[cite: 6]
    socket_fd = socket(sa.sin_family, SOCK_STREAM, 0);
    if (socket_fd == -1)
    {
        fprintf(stderr, "[Client %d] socket fd error: %s\n", client_id, strerror(errno));
        pthread_exit(NULL);
    }

    if (connect(socket_fd, (struct sockaddr *)&sa, sizeof sa) != 0)
    {
        fprintf(stderr, "[Client %d] connect error: %s\n", client_id, strerror(errno));
        close(socket_fd);
        pthread_exit(NULL);
    }

    printf("[Client %d] Connected to server %s:%d\n", client_id, server_ip, PORT);

    int num_messages = (rand() % 3) + 1;
    for (int i = 0; i < num_messages; i++)
    {
        char random_text[9];
        for (int j = 0; j < 8; j++)
        {
            random_text[j] = 'a' + (rand() % 26);
        }
        random_text[8] = '\0';

        char msg[256];
        sprintf(msg, "client %d msg %d: %s", client_id, i + 1, random_text);
        int msg_len = strlen(msg);

        // Send a message to server[cite: 6]
        if (send(socket_fd, msg, msg_len, 0) == -1)
        {
            fprintf(stderr, "[Client %d] send error\n", client_id);
            break;
        }

        // Wait for message from server via the socket[cite: 6]
        int bytes_read = recv(socket_fd, buffer, BUFSIZ - 1, 0);
        if (bytes_read > 0)
        {
            buffer[bytes_read] = '\0';
            printf("[Client %d] Response from server: \"%s\"\n", client_id, buffer);
        }
    }

    close(socket_fd);
    printf("[Client %d] Finished and closed socket\n", client_id);
    return NULL;
}

int main(void)
{
    printf("---- AUTOMATED MULTITHREADED CLIENT ----\n\n");
    srand(time(NULL));

    printf("Enter server IP address (e.g., 127.0.0.1): ");

    // NUEVO: Limpiamos cualquier salto de línea o espacio basura que Cygwin pueda meter
    if (fgets(server_ip, sizeof(server_ip), stdin) != NULL)
    {
        server_ip[strcspn(server_ip, "\r\n ")] = '\0';
    }

    pthread_t threads[NUM_CLIENTS];

    for (int i = 0; i < NUM_CLIENTS; i++)
    {
        thread_data_t *data = malloc(sizeof(thread_data_t));
        data->client_id = i + 1;

        if (pthread_create(&threads[i], NULL, client_thread, data) != 0)
        {
            fprintf(stderr, "Failed to create thread %d\n", i + 1);
        }
    }

    for (int i = 0; i < NUM_CLIENTS; i++)
    {
        pthread_join(threads[i], NULL);
    }

    printf("All clients finished.\n");
    return 0;
}
